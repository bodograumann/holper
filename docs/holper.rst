holper package
==============

Submodules
----------

holper.affine_seq module
~~~~~~~~~~~~~~~~~~~~~~~~

.. automodule:: holper.affine_seq
    :members:
    :undoc-members:
    :show-inheritance:


holper.core module
~~~~~~~~~~~~~~~~~~

.. automodule:: holper.core
    :members:
    :undoc-members:
    :show-inheritance:


holper.invoice module
~~~~~~~~~~~~~~~~~~~~~

.. automodule:: holper.invoice
    :members:
    :undoc-members:
    :show-inheritance:


holper.iofxml2 module
~~~~~~~~~~~~~~~~~~~~~

.. automodule:: holper.iofxml2
    :members:
    :undoc-members:
    :show-inheritance:


holper.iofxml3 module
~~~~~~~~~~~~~~~~~~~~~

.. automodule:: holper.iofxml3
    :members:
    :undoc-members:
    :show-inheritance:


holper.model module
~~~~~~~~~~~~~~~~~~~

.. automodule:: holper.model
    :members:
    :undoc-members:
    :show-inheritance:


holper.sportsoftware module
~~~~~~~~~~~~~~~~~~~~~~~~~~~

.. automodule:: holper.sportsoftware
    :members:
    :undoc-members:
    :show-inheritance:


holper.start module
~~~~~~~~~~~~~~~~~~~

.. automodule:: holper.start
    :members:
    :undoc-members:
    :show-inheritance:


holper.tools module
~~~~~~~~~~~~~~~~~~~

.. automodule:: holper.tools
    :members:
    :undoc-members:
    :show-inheritance:


Module contents
---------------

.. automodule:: holper
    :members:
    :undoc-members:
    :show-inheritance:
