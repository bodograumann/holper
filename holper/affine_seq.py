import re
from fractions import gcd

def lcm(a, b):
    return int(abs(a * b) / gcd(a,b)) if a and b else 0


# Affine sequences are described by strings of the form '5n+7', similar to nth-child in css
affine_sequence_re = re.compile(r'(?P<interval>[0-9]*)n(?:\+(?P<offset>[0-9]+))?')

class AffineSeq:
    __slots__ = ['start', 'stop', 'step']

    def __init__(self, start, stop, step = 1):
        if type(start) is str:
            match = affine_sequence_re.match(start)
            self.start = int(match.group('offset') or 0)
            self.step = int(match.group('interval') or 1)
        else:
            self.start = start
            self.step = step

        self.stop = stop

        assert(self.step > 0)

    def pretty(self):
        return (str(self.step) if self.step != 1 else '') + 'n' + ('+' + str(self.start) if self.start else '')

    def toRange(self):
        return range(self.start, self.stop, self.step)

    def __len__(self):
        return len(self.toRange())

    def __getitem__(self, key):
        if type(key) is AffineSeq:
            return AffineSeq(self[key.start], self[key.stop], self.step * key.step)
        else:
            return self.start + self.step * key

    def __iter__(self):
        return iter(self.toRange())

    def __reversed__(self):
        return reversed(self.toRange())

    def __contains__(self, item):
        return (self.start <= item < self.stop
                and (item - self.start) % self.step == 0)

    def __add__(self, other):
        return AffineSeq(self.start + other.start, self.stop + other.stop, self.step + other.step)

    def __lshift__(self, steps):
        return AffineSeq(self.start - steps * self.step, self.stop - steps * self.step, self.step)

    def __rshift__(self, steps):
        return AffineSeq(self.start + steps * self.step, self.stop + steps * self.step, self.step)

    def __and__(self, other):
        start = max(self.start, other.start)
        stop = min(self.stop, other.stop)
        step = lcm(self.step, other.step)

        if (self.start - other.start) % gcd(self.step, other.step):
            stop = start
        else:
            for item in self:
                if item in other:
                    start = item
                    break
            else:
                stop = start

        return AffineSeq(start, stop, step)

    def __repr__(self):
        return 'AffineSeq({}, {}, {})'.format(self.start, self.stop, self.step)

    def __str__(self):
        l = len(self)
        s = '('
        if l > 0:
            s += str(self.start)
        if l > 1:
            s += ', ' + str(self.start + self.step)
        if l > 3:
            s += ', â€¦'
        if l > 2:
            s += ', ' + str(next(reversed(self)))
        s += ')'

        return s
