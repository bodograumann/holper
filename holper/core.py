"""Core functionality"""

from itertools import groupby
from typing import Callable, Optional
import logging
import sqlalchemy
from . import tools, model

_logger = logging.getLogger(__name__)


def open_session(source: str) -> sqlalchemy.orm.Session:
    engine = sqlalchemy.create_engine(source, echo=False)
    if source.startswith("sqlite:"):
        tools.fix_sqlite_engine(engine)
    model.Base.metadata.create_all(engine)

    Session = sqlalchemy.orm.sessionmaker(bind=engine)
    return Session(future=True)


def get_event(session: sqlalchemy.orm.Session, event_id: int) -> Optional[model.Event]:
    try:
        (event,) = next(session.execute(sqlalchemy.select(model.Event).where(model.Event.event_id == event_id)))
        return event
    except StopIteration:
        return None


def get_race(session: sqlalchemy.orm.Session, race_id: int) -> Optional[model.Race]:
    try:
        (race,) = next(session.execute(sqlalchemy.select(model.Race).where(model.Race.race_id == race_id)))
        return race
    except StopIteration:
        return None


def hydrate_country_by_ioc_code(session, entity):
    if not entity or not entity.country:
        return

    ioc_code = entity.country.ioc_code
    result = session.execute(sqlalchemy.select(model.Country).where(model.Country.ioc_code == ioc_code))
    try:
        (country,) = next(result)
    except StopIteration:
        _logger.warning("Could not find country with ioc_code “%s” — Clearing.", ioc_code)
        entity.country = None
        return

    entity.country = country


def shadow_entity_by_xid(session, entity):
    cls = entity.__class__
    xid_cls = getattr(model, cls.__name__ + "XID")
    for xid in entity.external_ids:
        result = session.execute(
            sqlalchemy.select(xid_cls).where(xid_cls.issuer == xid.issuer).where(xid_cls.external_id == xid.external_id)
        )
        try:
            (saved_xid,) = next(result)
        except StopIteration:
            continue
        return getattr(saved_xid, tools.camelcase_to_snakecase(cls.__name__))
    return entity


def group_courses_by_first_control(race: model.Race) -> dict[str, list[model.Course]]:
    get_first_control: Callable[[model.Course], str] = lambda course: course.controls[1].control.label

    return {
        control_label: list(course_group)
        for control_label, course_group in groupby(sorted(race.courses, key=get_first_control), get_first_control)
    }
