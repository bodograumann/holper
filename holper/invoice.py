"""Classes to manage invoices"""

import datetime
import decimal

from collections import OrderedDict

import pystache

precision = decimal.Decimal("0.01")


def round_amount(number):
    return number.quantize(precision)


class Prices:
    def __init__(self):
        self.items = OrderedDict()

    def add_price(self, item_id, price, description, group=None):
        self.items[item_id] = (round_amount(decimal.Decimal(price)), description, group)

    def get_price(self, item_id):
        return self.items[item_id][0]

    def get_description(self, item_id):
        return self.items[item_id][1]

    def get_group_name(self, item_id):
        return self.items[item_id][2]


class Invoice:
    def __init__(self, prices, recipient, date_format="%y-%m-%d"):
        self.prices = prices
        self.recipient = recipient
        self.date_format = date_format
        self.groups = OrderedDict()
        self.paid_amount = 0
        self.remark = ""

    def add_items(self, item_id, amount=1):
        if not amount:
            return

        group = self.prices.get_group_name(item_id)
        if group not in self.groups:
            self.groups[group] = {}

        price = self.prices.get_price(item_id)
        if item_id in self.groups[group]:
            self.groups[group][item_id]["amount"] += amount
            self.groups[group][item_id]["total"] += round(price * amount)
        else:
            self.groups[group][item_id] = {
                "description": self.prices.get_description(item_id),
                "price": price,
                "amount": amount,
                "total": price * amount,
            }

    def set_paid(self, amount):
        self.paid_amount = decimal.Decimal(amount)

    def set_remark(self, remark):
        self.remark = remark

    def get_total(self):
        return sum(item["total"] for group in self.groups.values() for item in group.values())

    def check_total(self, total):
        return round(decimal.Decimal(total)) == round(self.get_total())

    def fill_template(self, template_file, target_file, labels, reserve_space=False):
        subtotal = self.get_total()
        data = {
            "date": datetime.date.today().strftime(self.date_format),
            "recipient": self.recipient,
            "labels": labels,
            "groups": [
                {
                    "group_name": group,
                    "items": [
                        self.groups[group][item_id] for item_id in self.prices.items if item_id in self.groups[group]
                    ],
                }
                for group in self.groups
            ],
            "subtotal": round(subtotal),
            "paid": round(self.paid_amount),
            "total": round(subtotal - self.paid_amount),
            "reserve_space": reserve_space,
            "remark": self.remark,
        }

        renderer = pystache.Renderer()
        template = renderer.load_template(template_file)
        invoice = renderer.render(template, data)

        with open(target_file, "w") as target:
            target.write(invoice)
