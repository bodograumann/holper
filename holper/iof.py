from datetime import datetime
from itertools import chain

from pydantic_xml import BaseXmlModel, attr, element, wrapped
# Unsupported:
# - forward refs with `from future import annotated`

nsmap = {
    "": "http://www.orienteering.org/datastandard/3.0",
    "xsi": "http://www.w3.org/2001/XMLSchema-instance",
}


class IOFBaseModel(BaseXmlModel, nsmap=nsmap, search_mode="ordered"):
    pass


class ClassCourseAssignment(IOFBaseModel):
    class_name: str = element(tag="ClassName")
    course_name: str = element(tag="CourseName")


class TeamMemberCourseAssignment(IOFBaseModel):
    leg: int = element(tag="Leg")
    course_name: str = element(tag="CourseName")
    course_family: str = element(tag="CourseFamily")


class TeamCourseAssignment(IOFBaseModel):
    bib_number: str = element(tag="BibNumber")
    team_member_course_assignments: list[TeamMemberCourseAssignment] = element(tag="TeamMemberCourseAssignment")


class CourseControl(IOFBaseModel):
    type: str = attr()
    control: str = element(tag="Control")
    leg_length: int | None = element(tag="LegLength")


class Course(IOFBaseModel):
    name: str = element(tag="Name")
    course_family: str | None = element(tag="CourseFamily")
    length: int = element(tag="Length")
    climb: int = element(tag="Climb")
    course_controls: list[CourseControl] | None = element(tag="CourseControl")


class MapPosition(IOFBaseModel):
    x: float = attr()
    y: float = attr()
    unit: str = attr()


class Control(IOFBaseModel):
    control_id: str = element(tag="Id")
    map_position: MapPosition = element(tag="MapPosition")


class Map(IOFBaseModel):
    scale: int = element(tag="Scale")


class RaceCourseData(IOFBaseModel):
    map: Map = element(tag="Map")
    controls: list[Control] = element(tag="Control")
    courses: list[Course] = element(tag="Course")
    team_course_assignments: list[TeamCourseAssignment] | None = element(tag="TeamCourseAssignment")
    class_course_assignments: list[ClassCourseAssignment] | None = element(tag="ClassCourseAssignment")

    def delete_unused_courses(self):
        used_courses = [
            assignment.course_name
            for assignment in chain(
                chain.from_iterable(
                    team_assignment.team_member_course_assignments for team_assignment in self.team_course_assignments
                ),
                self.class_course_assignments,
            )
        ]
        self.courses = [course for course in self.courses if course.name in used_courses]


class Event(IOFBaseModel):
    name: str = element(tag="Name")


class CourseData(IOFBaseModel):
    iof_version: str = attr(name="iofVersion")
    create_time: datetime = attr(name="createTime")
    creator: str = attr()

    event: Event = element(tag="Event")
    race_course_data: RaceCourseData = element(tag="RaceCourseData")
