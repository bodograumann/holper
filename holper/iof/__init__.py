"""Representation of the IOF Data Standard, version 3.0 as pydantic-xml models.

We cannot parse Extensions, because they might contain arbitrary data.

pydantic-xml does not support forward refs with `from __future__ import annotations`, so we define all models in roughly inverted order from the xml schema.

Cf. https://github.com/international-orienteering-federation/datastandard-v3/blob/master/IOF.xsd
"""

from .class_list import ClassList
from .competitor_list import CompetitorList
from .control_card_list import ControlCardList
from .course_data import CourseData
from .entry_list import EntryList
from .event_list import EventList
from .organisation_list import OrganisationList
from .result_list import ResultList
from .service_request_list import ServiceRequestList
from .start_list import StartList

__all__ = [
    "ClassList",
    "CompetitorList",
    "ControlCardList",
    "CourseData",
    "EntryList",
    "EventList",
    "OrganisationList",
    "ResultList",
    "ServiceRequestList",
    "StartList",
]
