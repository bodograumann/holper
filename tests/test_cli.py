from importlib.resources import as_file, files
from tempfile import NamedTemporaryFile

import pytest
from typer.testing import CliRunner

from holper import cli

from . import events

runner = CliRunner()


@pytest.fixture()
def db_file_opt():
    with NamedTemporaryFile() as db_file:
        yield ("--db-file", db_file.name)


@pytest.fixture()
def cli_snapshot(db_file_opt, snapshot):
    def _cli_snapshot(*args, exit_code=0):
        result = runner.invoke(cli.app, [*db_file_opt, *args])
        assert result.exit_code == exit_code, f"Unexpected exit code. Output: \n{result.stdout}"
        assert result.stdout == snapshot
        return result

    return _cli_snapshot


def test_create_event(cli_snapshot):
    cli_snapshot("init")
    cli_snapshot("new-event", "event.test-name.1", "2024-02-29 10:00:00")
    cli_snapshot("events")


@pytest.mark.parametrize("event", files(events).iterdir())
def test_imports_and_startlist(event, cli_snapshot):
    with as_file(event) as event_dir:
        cli_snapshot("init")
        cli_snapshot("new-event", event.name, "2024-02-29 10:00:00")
        cli_snapshot("import-categories", "1", str(event_dir / "ClassList.xml"))
        cli_snapshot("import-courses", "-s", "1", str(event_dir / "CourseData.xml"))
        cli_snapshot("import-entries", "1", str(event_dir / "EntryList.xml"))

        cli_snapshot("events")
        cli_snapshot("event", "1")
        cli_snapshot("entries", "1")

        cli_snapshot("startlist", "1", "3", "--parallel-max", "6", "--greedy")
