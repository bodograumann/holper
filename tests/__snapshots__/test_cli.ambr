# serializer version: 1
# name: test_create_event
  '''
  A new event “event.test-name.1” was created successfully with number #1.
  
  '''
# ---
# name: test_create_event.1
  '''
  #1 2024-02-29 10:00:00 - 2024-03-01 10:00:00: event.test-name.1, 1 races, 0 categories, 0 entries
  
  '''
# ---
# name: test_imports_and_startlist[event0]
  '''
  A new event “HerbstOL2022” was created successfully with number #1.
  
  '''
# ---
# name: test_imports_and_startlist[event0].1
  '''
  Imported 36 categories
  
  '''
# ---
# name: test_imports_and_startlist[event0].2
  '''
  Imported 16 courses
  
  '''
# ---
# name: test_imports_and_startlist[event0].3
  '''
  Imported 157 entries
  
  '''
# ---
# name: test_imports_and_startlist[event0].4
  '''
  #1 2024-02-29 10:00:00 - 2024-03-01 10:00:00: HerbstOL2022, 1 races, 36 categories, 157 entries
  
  '''
# ---
# name: test_imports_and_startlist[event0].5
  '''
  Name: HerbstOL2022
  Period: 2024-02-29 10:00:00 - 2024-03-01 10:00:00
  Form: individual
  1 Races: 2024-02-29 10:00:00
  36 Categories: BK, BL, D-10, D-12, D-14, D-16, D-18, D19K, D19L, D35-, D40-, D45-, D50-, D55-, D60-, D65-, D70-, D75-, D80-, H-10, H-12, H-14, H-16, H-18, H19K, H19L, H35-, H40-, H45-, H50-, H55-, H60-, H65-, H70-, H75-, H80-
  157 Entries
  
  '''
# ---
# name: test_imports_and_startlist[event0].6
  '''
  Katja Lahm: D50-
  Marlene Ullmann: D-12
  Lina Zumbühl: D-14
  Wolfgang Kössler: H65-
  Erika Polgar: D19K
  Anatol Zeh: H-14
  Christian Kremer: H70-
  Laila Poschitzki: D-16
  Emma Poschitzki: D-16
  Janik Keller: H-16
  Sibylle Uhlemann: D45-
  Heidrun Graumann: D55-
  Sabine Richter: D55-
  Bernd Graumann: H60-
  Felicitas Vogt: D19L
  Kathrin Trappe: D35-
  Linda Brandt: D-16
  Katrin Brandt: D40-
  Jana Storch: D45-
  Andreas Neitzel: H19K
  Finjas Cain: H-10
  Thomas Krieger: H45-
  Stefan Röhl: H50-
  Andreas Mevius: H60-
  Oskar Frenzel: H19L
  Friederike Frenzel: D19L
  Vera Lotte Adorján: D-12
  Petru Tulban: H-16
  Leopold Kühschelm: H40-
  Péter Adorján: H40-
  Rudolf Herold: H70-
  Nelio Gergs: H-16
  Frank Mieske: H35-
  Christian Appelt: H40-
  Katrin Kubald: D50-
  Henry Jobst: H50-
  Uwe Brettschneider: H50-
  Gudrun Feldmann: D55-
  Anne Ferchow: D55-
  Sven Demmig: H55-
  Bernd Käding: H55-
  Gerhard Brettschneider: H55-
  André Feldmann: H55-
  Thomas Grassow: H55-
  Jochen Winkler: H55-
  Gudrun Brettschneider: D60-
  Petra Grassow: D60-
  Hartmut Brettschneider: H60-
  Gabriela Lehmann: D65-
  Kerstin Mielecke: D65-
  Norbert Pommer: H65-
  Sascha Boeck: H19L
  Margarita Dimova: D19L
  Carola Schipke: D19K
  Erwin Schramm: H19K
  Ralph Yeung: H19K
  Dennis Bräuer: H19L
  Christian Kehlmaier: H50-
  Justus Sperling: H-10
  Maxim Hirschmann: H-14
  Lucas Pascal Mache: H-14
  Gerhard Plötz: H70-
  David Knuth: BK
  Heiko Sperling: BL
  Jacqueline Siebert: BL
  Ella Sperling: D-10
  Kamilla Knuth: D-10
  Gergely Knuth: H-16
  Sara Knuth: D-18
  Fidelius Cain: H-18
  Juliane Kummerlöwe: D35-
  Max Metzger: H35-
  Anna Dost: D45-
  Johanna Söter: D45-
  Tinko Siebert: H45-
  Frank-Dieter Hoffmann: H55-
  Martin Füßling: D65-
  Michael Großmann: H19L
  Kristina Bringezu: D19L
  Kati Tschischkale: D19K
  Thomas Mache: H19K
  Marcin GRYDZIUSZKO: H45-
  Mariusz KRAMSKI: H45-
  Krzysztof PACYNA: H45-
  Jerzy Deptuła: H55-
  Wojciech MELIŃSKI: H65-
  Ralf Schwanebeck: BK
  Martina Fink: D45-
  Wolfgang Fink: H65-
  Paul Köhler: H-14
  Gunnar Schälicke: H-14
  Nicolae Tulban: H50-
  Ivanna Nedoborovsky: D-10
  Daniel Nedoborovsky: H-18
  Andrey Nedoborovsky: H50-
  Matti Alonen: H19L
  Bernd Wollenberg: H70-
  Diana Chowdhury: D40-
  Bela Winnig: H-10
  Ida Leukert: D-12
  Susi Thiel: D75-
  Marlene Jürgens: BK
  Susann Rohloff: BK
  Oleg Kharitonov: BL
  Aksana Kharitonova: BL
  Jörg Pliquett: BL
  Linda Lange: D-10
  Maksim Kharitonov: H-12
  Lina Leukert: D-16
  Maria Tovchigrechko: D-18
  Jonas Lange: H-18
  Anke Winnig: D45-
  Alina Shmakova: D45-
  Stefan Edelhoff: H45-
  Stefan Winterfeldt: H45-
  Uwe Röhner: H45-
  Annette Leonhardt: D50-
  Kristin Leukert: D50-
  Matthias Forberg: H50-
  Udo Lange: H50-
  Frank Braatz: H50-
  Kerstin Ahlburg: D55-
  Kai Dalüge: H55-
  Rainer Ahlburg: H55-
  Monika Braatz: D65-
  Uwe Jürgens: H65-
  Burkhard Jandrig: H65-
  Denis Winnig: H19L
  Raik Zschäckel: H19L
  Robin Zschäckel: H19L
  Anke Xylander: D19L
  Imanuel Fröhlich: H-14
  Hans-Peter Blome: H65-
  Hannes Fröhlich: H19K
  Stine Friederike Siebert: D19L
  Andriy Martsyniv: H45-
  Jakob und Bosse Hallier und Bahrenberg: H-10
  Valentin und Carla Hallier und Bahrenberg: H-10
  Artur und Leo Rahn: H-10
  Beate Dalüge: BK
  Franziska Bonewitz: BK
  Alina Orlanski: BK
  Paul Orlanski: BK
  Anna Orlianskaia: BK
  Olga Levitski: BK
  Michael Herms: BL
  Anna borodina: BL
  Daniil Levitski: BL
  Anne Jürgens: D35-
  Yevhen Smokov: H19L
  Juliane Lenter: D19L
  Maryna Yudina: D19L
  Alena Yurpalova: D19K
  Ralf Bahrenberg: H19K
  Tobias Christ: H19K
  Anton Yurpalov: H19K
  Dimitry borodin: H19K
  
  '''
# ---
# name: test_imports_and_startlist[event1]
  '''
  A new event “WinterOL2023” was created successfully with number #1.
  
  '''
# ---
# name: test_imports_and_startlist[event1].1
  '''
  Imported 35 categories
  
  '''
# ---
# name: test_imports_and_startlist[event1].2
  '''
  Could not find category Leo.
  Could not find category Andreas.
  Could not find category Jürgen.
  Could not find category Tamas.
  Imported 18 courses
  
  '''
# ---
# name: test_imports_and_startlist[event1].3
  '''
  Imported 156 entries
  
  '''
# ---
# name: test_imports_and_startlist[event1].4
  '''
  #1 2024-02-29 10:00:00 - 2024-03-01 10:00:00: WinterOL2023, 1 races, 35 categories, 156 entries
  
  '''
# ---
# name: test_imports_and_startlist[event1].5
  '''
  Name: WinterOL2023
  Period: 2024-02-29 10:00:00 - 2024-03-01 10:00:00
  Form: individual
  1 Races: 2024-02-29 10:00:00
  35 Categories: BK, BL, D-10, D-12, D-14, D-16, D-18, D19-K, D19-L, D35-, D40-, D45-, D50-, D55-, D60-, D65-, D70-, D75-, H-10, H-12, H-14, H-16, H-18, H19-K, H19-L, H35-, H40-, H45-, H50-, H55-, H60-, H65-, H70-, H75-, H80-
  156 Entries
  
  '''
# ---
# name: test_imports_and_startlist[event1].6
  '''
  Katja Lahm: D55-
  Wolfgang Kössler: H65-
  Erika Polgar: D19-K
  Emilia Schlittermann: D-12
  Mariella Kleiner: D-12
  Bernd Graumann: BK
  Emma Reinhardt: D-10
  Janik Keller: H-18
  Heidrun Graumann: D55-
  Michael Neidhardt: H60-
  Egon Johannes Trappe: H-10
  Kathrin Trappe: D35-
  Frank Hertel: BL
  Adrian Hertel: H-18
  Katrin Brandt: D45-
  Katrin Fillinger: D50-
  Andreas Neitzel: H50-
  Dirk Mönch: H55-
  Ulrike Neitzel: D19-L
  Lutz Spranger: H35-
  Thomas Krieger: H45-
  Michael Hase: H55-
  Holger Farrak: H50-
  Linnéa Hanßke: BK
  Lotta Hanßke: BK
  Lena Höhne: D19-L
  Lene Powik: D-14
  Alina Berens: D-14
  Paula Weiß: D-14
  Marla Powik: D-10
  Amelie Knaack: D-10
  Anton Weiß: H-12
  Ronja Fell: D75-
  Ivonne Jobst: BK
  Thessa Jobst: BL
  Lionel Kumbier: H-16
  Carl Costa Green: H-16
  Petru Tulban: H-18
  Tobias Fell: H35-
  Matthias Schwartz: H40-
  Péter Adorján: H45-
  Barbara Usemann: D70-
  Madeline Gericke: D70-
  Rudolf Herold: H70-
  Silke Herold: BK
  Tobias Licht: BL
  Paulina /Max Brettschneider: D-10
  Moritz Brettschneider: H35-
  Christian Appelt: H45-
  Felix Mihram: H45-
  Jochen Winkler: H45-
  Katrin Kubald: D50-
  Henry Jobst: H50-
  Anne Ferchow: D55-
  Kirsten Käding: D55-
  Sven Demmig: H55-
  Gerhard Brettschneider: H55-
  Bernd Käding: H55-
  Gudrun Brettschneider: D60-
  Kerstin Mielecke: D65-
  Gabriela Lehmann: D65-
  Norbert Pommer: H65-
  Konstanze Gergs: D19-K
  Sascha Boeck: H19-L
  Marlen Manns: D-12
  Frida Gühne: D-12
  Heiko Sperling: BL
  Ella Sperling: D-10
  Mia Rex: D-10
  Clara Joppich: D-10
  Justus Sperling: H-12
  Finjas Cain: H-12
  Leander Fischer: H-16
  Livia Uhlemann: D-18
  Fidelius Cain: H-18
  Henryk Dobslaw: H35-
  Yvonne Fischer: D40-
  Martin Füßling: H40-
  Tinko Siebert: H45-
  Matthias Wauer: H65-
  Kati Tschischkale: D19-K
  Michael Großmann: H19-L
  Kristina Bringezu: D19-L
  Nora von Szada: D19-L
  Heinrich Schwanebeck: H-14
  Wolfgang Fink: H70-
  Ralf Schwanebeck: BK
  Theodor Schwanebeck: H-16
  Martina Fink: D45-
  Ole Engel: H-10
  Karl Herold: H-10
  Emil Schmidt: H-10
  Hannah Kalisch: D-12
  Ella Schmidt: D-12
  Johanna Burczyk: D-12
  Tjare Funke: H-14
  Louisa Kalisch: D-10
  Paul Köhler: H-16
  Gunnar Schälicke: H-16
  Olaf Schmidt: H45-
  Katrin Herold: D19-K
  Fabian Kalisch: H19-K
  Nicolae Tulban: H50-
  Alexander Büchner: H19-K
  Bernd Wollenberg: H75-
  Lina Peukert: BK
  Michael Peukert: BK
  Juan Francisco Roggerone: H19-K
  Dmitry Ushakov: H19-L
  Ida Leukert: D-12
  Rolf Thiel: H80-
  Susi Thiel: BK
  Bernd Schüler: BL
  Maksim Kharitonov: H-12
  Lina Leukert: D-16
  Rafael Schueler: H-16
  Maria Tovchigrechko: D-18
  Raik Zschäckel: H35-
  Virginijus Purvys: H35-
  Aksana Kharitonova: D40-
  Oleg Kharitonov: H40-
  Anke Winnig: D45-
  Bettina Rudert: D50-
  Daniela Zschäckel: D50-
  Kristin Leukert: D50-
  Andreas Rudert: H50-
  Stefan Bleidorn: H50-
  Matthias Forberg: H50-
  Stefan Edelhoff: H50-
  Annette Leonhardt: D55-
  Kerstin Ahlburg: D55-
  Rainer Ahlburg: H55-
  Kai Dalüge: H55-
  Frank Braatz: H60-
  Monika Braatz: D65-
  Uwe Jürgens: H65-
  Robin Zschäckel: H19-L
  Denis Winnig: H19-L
  Matthias Kahlert: H55-
  Urs Trösch: H55-
  Hannes Fröhlich: H19-K
  Karsten Blume: H19-L
  Andriy Martsyniv: H45-
  Michael Herms: BK
  Doreen Köhler: BK
  Ulrike Engel: BK
  Beate Dalüge: BK
  Sandra Schöb-Dirks: BK
  Anna / Carla Bleidorn / Sedlin: D-10
  Benjamin Roggendorf: H-18
  Valerii Prosolkov: H40-
  Stefan Dirks: H45-
  Matthias Roggendorf: H45-
  Yevhen Smokov: H19-K
  Johannes Heizmann: H19-K
  Juliane Lenter: D19-L
  
  '''
# ---
# name: test_imports_and_startlist[event2]
  '''
  A new event “BBMSprint2022” was created successfully with number #1.
  
  '''
# ---
# name: test_imports_and_startlist[event2].1
  '''
  Imported 24 categories
  
  '''
# ---
# name: test_imports_and_startlist[event2].2
  '''
  Could not find category Leo.
  Could not find category Jürgen.
  Could not find category Tamas.
  Could not find category Christian.
  Could not find category Andreas.
  Could not find category Johannes.
  Could not find category Bernd.
  Could not find category Michael.
  Imported 21 courses
  
  '''
# ---
# name: test_imports_and_startlist[event2].3
  '''
  Imported 135 entries
  
  '''
# ---
# name: test_imports_and_startlist[event2].4
  '''
  #1 2024-02-29 10:00:00 - 2024-03-01 10:00:00: BBMSprint2022, 1 races, 24 categories, 135 entries
  
  '''
# ---
# name: test_imports_and_startlist[event2].5
  '''
  Name: BBMSprint2022
  Period: 2024-02-29 10:00:00 - 2024-03-01 10:00:00
  Form: individual
  1 Races: 2024-02-29 10:00:00
  24 Categories: BK, BL, D-10, D-12, D-14, D-16, D-18, D19-, D35-, D45-, D55-, D65-, D75-, H-10, H-12, H-14, H-16, H-18, H19-, H35-, H45-, H55-, H65-, H75-
  135 Entries
  
  '''
# ---
# name: test_imports_and_startlist[event2].6
  '''
  Erika Polgar: D45-
  Wolfgang Kössler: H65-
  Anatol Zeh: H-14
  Janik Keller: H-16
  Heidrun Graumann: D55-
  Lutz Wohlrabe: H55-
  Friederike Graumann: D19-
  Sabine Richter: D55-
  Kathrin Trappe: D35-
  Frank-Peter Frenzel: H55-
  Maren Brandt: BK
  Linda Brandt: D-16
  Katrin Brandt: D35-
  Andreas Lau: H45-
  Andreas Neitzel: H45-
  Ulrike Neitzel: D19-
  Martin Röhl: H45-
  Britta Brünings: D55-
  Dieter Müller-Danert: H55-
  Henning Kapischke: H19-
  Vera Lotte Adorján: D-12
  Clara Schulz: D-12
  Lionel Kumbier: H-14
  Andrea Naß: BK
  Thea Schulz: D-10
  Frida Magurno: D-10
  Serafina Quabeck: D-10
  Franziska Lau: D-16
  Petru Tulban: H-16
  Claudia Becker: D35-
  Noemi Quabeck: D35-
  Leopold Kühschelm: H35-
  Matthias Schwartz: H35-
  Pina Mauch: D19-
  Sebastian Fleiß: H19-
  Linus Appelt: H-14
  Nelio Gergs: H-14
  Erika Lemnitzer: D75-
  Eberhard Lemnitzer: H75-
  Sascha Boeck: BK
  Konstanze Gergs: BL
  Max Brettschneider: BL
  Eva Heilmann: BL
  Ulrike Thieme: BL
  Katrin Herold: D35-
  Christian Appelt: H35-
  Erwin Schramm: H35-
  Robert Jäschke: H35-
  Katrin Kubald: D45-
  Carola Schipke: D45-
  Henry Jobst: H45-
  Gudrun Brettschneider: D55-
  Kirsten Käding: D55-
  André Feldmann: H55-
  Sven Demmig: H55-
  Bernd Käding: H55-
  Thomas Adam: H55-
  Gabriela Lehmann: D65-
  Rudolf Herold: H65-
  Gerhard Brettschneider: H65-
  Jochen Winkler: H65-
  Margarita Dimova: D19-
  Henriette Käding: D19-
  Isabell Jänich: D19-
  Matthias Müller: H55-
  Justus Sperling: H-10
  Ella Sperling: BK
  Dorian Dost: BK
  Kamilla Knuth: D-10
  Gergely Knuth: H-16
  Leander Fischer: H-16
  Sara Knuth: D-18
  Yvonne Fischer: D35-
  Martin Füßling: H35-
  Kristina Bringezu: D45-
  David Knuth: H45-
  Kati Tschischkale: D55-
  Lina Peukert: D-10
  Karl Herold: H-10
  Paul Köhler: H-14
  Gunnar Schälicke: H-14
  Tjare Funke: H-14
  Ella Schmidt: D-14
  Andreea Tulban: BL
  Nicolae Tulban: H45-
  Sabine Schramm: BK
  Peter Petrov: H45-
  Bernd Wollenberg: H65-
  Ida Leukert: D-12
  Susi Thiel: D75-
  Rolf Thiel: H75-
  Helmut Fillinger: H75-
  Yves Baujard: H75-
  Irina Wollmerstädt: BL
  Monika Braatz: BL
  Sascha Schulz: BL
  Linda Lange: D-10
  Maksim Kharitonov: H-12
  Lina Leukert: D-16
  Maria Tovchigrechko: D-16
  Rafael Schueler: H-16
  Jonas Lange: H-18
  Aksana Kharitonova: D35-
  Oleg Kharitonov: H35-
  Kristin Leukert: D45-
  Alina Shmakova: D45-
  Bettina Rudert: D45-
  Matthias Forberg: H45-
  Udo Lange: H45-
  Frank Braatz: H45-
  Annette Leonhardt: D55-
  Kai Dalüge: H55-
  Lea Luttenberger: D19-
  Daniela Zschäckel: D19-
  Stefan Bleidorn: H19-
  Raik Zschäckel: H19-
  Robin Zschäckel: H19-
  Uwe Röhner: H19-
  Katerina Wollmerstädt: D45-
  Anke Templiner: D45-
  Jeannette Molnar: D55-
  Galina Scholtz: D55-
  Urs Trösch: H55-
  Hans-Peter Blome: H65-
  Aurelien Filiali: H19-
  Karsten Blume: H19-
  Sandra Schöb-Dirks: BK
  Neo Lau: BK
  Beate Dalüge: BL
  Franziska Bonewitz: BL
  Stefan Dirks: H45-
  Olaf Schmidt: H45-
  Kseniya Dabrytskaya: D19-
  Alena Yurpalova: D19-
  Anton Yurpalov: H19-
  
  '''
# ---
